pennylane/gradients/general_shift_rules.py

```diff
@@ -215,33 +207,20 @@ def _iterate_shift_rule(rule, order, period=None):
     if len(rule[0]) == 3:
         return _iterate_shift_rule_with_multipliers(rule, order, period)
 
-    combined_rules = []
-
-    # TODO: Optimization: Only iterate over half of the combinations and
-    # double their coefficients
-    for partial_rule in itertools.product(rule, repeat=order):
-        c, s = np.stack(partial_rule).T
-        cumul_shift = sum(s)
-        if period is not None:
-            cumul_shift = np.mod(cumul_shift + 0.5 * period, period) - 0.5 * period
-        combined_rules.append(np.stack([np.prod(c), cumul_shift]))
-
-    # combine all terms in the linear combination into a single
-    # array, with coefficients on the first column and shifts on the second column.
-    return qml.math.stack(combined_rules)
-
-
-def _combine_shift_rules_with_multipliers(rules):
-    r"""Helper method to combine shift rules for multiple parameters that
-    contain multipliers into simultaneous multivariate shift rules."""
-    combined_rules = []
-
-    for partial_rules in itertools.product(*rules):
-        c, m, s = np.stack(partial_rules).T
-        combined = np.concatenate([[np.prod(c)], m, s])
-        combined_rules.append(np.stack(combined))
+    # TODO: optimization: Without multipliers, the order of shifts does not matter,
+    # so that we can only iterate over the symmetric part of the combined_rules tensor.
+    # This requires the corresponding multinomial prefactors to be factored into the coeffs.
 
-    return np.stack(combined_rules).T
+    # the following might become slow if high-order derivatives or very large rules are used
+    combined_rules = np.array(list(itertools.product(rule, repeat=order)))

```

Comment:
Do you mean if order >> 1?

Reply:
Yes, because there will be many terms in `combined_rules` On the other hand, the old code produced a list of the same size, so maybe this comment is a bit too pessimistic even? :thinking:

Reply:
(I removed this comment, because I don't think it actually adds value/information)